_posts/2022/2022-07-25-owning-your-cloud.md

```diff
@@ -0,0 +1,63 @@
+
+The first "abstraction" to break was relying on Azure Bot Service. It provides a single API to make it possible to write a bot once that can communicate with multiple chat services. Where have we heard that promise before, right? As you might expect, it meant supporting the lowest common denominator of what chat platforms have to offer. And this worked for our initial product idea, [ChatOps as a service](https://news.ycombinator.com/item?id=27974077). With ChatOps, the interface was primarily textual, so we didn't need all the bell and whistles. But as we started pivoting to providing a Customer Success product to help keep on top of conversations in Slack, it became imperative that we supported all features that Slack had to offer.
+
+This meant bypassing Bot Service and writing our code to directly interact with the Slack API. I don't regret that we started with Bot Service as it got us far fast, but I do regret when we configured our Slack Events Endpoints (the URLs that Slack sends its events to), we configured it with the Bot Service URLs. Something like `https://slack.botframework.com/api/messages/events` (or whatever it is). The reason this was a problem is that we submitted our Slack App to the Slack App Directory. So every change we make to the app's configuration can take up to six weeks to be approved by Slack. This meant that even though we were ready to go with our new approach, we had to wait for the Slack approval process to start serving Slack requests directly.
+
+## Lesson: Own Your Traffic With A Level of Indirection built in
+
+Ideally, we would have set up a DNS entry and configured Slack with that. It would have allowed us to point it to Azure Bot Service in the beginning, and later point it to our own web server when we needed to switch. At the time, we felt we wouldn't need it. But it's inexpensive to set up a DNS entry for a level of indirection.
+
+As a corollary to this, it's really helpful to manage DNS as code. We have several DNS providers, but we now manage all our DNS in a repository using https://github.com/octodns.
+
+Alright, back to the story. So what do you think happened once the approval went through and we were directing Slack traffic directly to our server? Of course, we found a bug and it didn't work. And we made the mistake of pointing Slack directly to https://ab.bot/ which is our main site, so we couldn't just point DNS back to botframework.com. We were really in a bind because it would take another six weeks to change the Slack configuration back to pointing to Bot Service.
+
+## Lesson: Own your Ingress
+
+Yes, we should have used a DNS entry other than our main app's entry. In retrospect it's so obvious. But even so, we also should have had some sort of forward proxy in place such as Azure Front Door or Azure Application Gateway. I should mention, neither of those choices worked out for us, which leads me to the next lesson.
```

Comment:
Essentially, Front Door and Application Gateway together supported the features we needed, but each one was missing something. It felt unnecessary (in the end) to set it up. We don't have any kind of front-door ingress as of now.